### PR TITLE
Prealloc for jectcoeffs!

### PR DESCRIPTION
**Pull request by lbenet**

### CODE REVIEW

**Comment by lbenet** [comment]
This PR proposes a new `@taylorize`, which creates two methods (instead of one): an internal preallocation function (`_allocate_jetcoeffs!`), that allocates all auxiliary `Taylor1` objects and vectors of `Taylor1` needed, and the (usual) improved `jetcoeffs!` method, which uses the preallocated arrays. The idea is that by allocating these objects *once* and reusing them, the number allocations and then the running time are reduced.

The following are very preliminary benchmarks, using the [`kepler4!` ODEs](https://github.com/PerezHz/TaylorIntegration.jl/blob/5567f757358d2e906dace08308220a86ff4c059f/test/taylorize.jl#L663) and `tf = 10_000.0`. The quantity I'd like to focus are the number of allocations:

In current master:
- parse_eqs=false: 8.017 s (57010139 allocations: 7.84 GiB)
- parse_eqs=true:  844.397 ms (3650983 allocations: 906.38 MiB)

This PR:
- parse_eqs=false: 7.924 s (57010142 allocations: 7.84 GiB)
- parse_eqs=true:  573.154 ms (702226 allocations: 315.02 MiB)

Considering even longer integration times should show even better impact in memory and running time.

**Comment by lbenet** [comment]
Rebased to current master (after #138 was merged). Tests are passing, but coverage has been slightly reduced.

I think this is ready for review @PerezHz; I'll bump the minor version once this is ready to be merged.

**Comment by lbenet** [comment]
Rebasing to current master (after #134 was merged).

**Comment by lbenet** [comment]
Grrrrr.... I'm not sure what's going on with the tests. Locally, tests pass both in Julia 1.7.2 and 1.8.0-beta3. I guess it is related to upgrades in DiffEqBase, OrdinaryDiffEq and/or RecursiveArrayTools.

**Comment by lbenet** [comment]
@PerezHz I fixed a few pending details. I think this is ready for review now. Once I have your comments, I can then bump the minor version

File: test/TestPkg/src/TestPkg.jl [hunk 1]
```diff
@@ -129,30 +129,30 @@ function TaylorIntegration.jetcoeffs!(::Val{f2!}, t::Taylor1{_T}, q::AbstractVec
...
-local μ = params
-X = Array{_eltype_q_}(undef, N, N)
-accX = Array{_eltype_q_}(undef, N) #acceleration
-for j in 1:N
-    accX[j] = zero(q[1])
-    dq[j] = q[N+j]
-end
-#compute accelerations
-for j in 1:N
+    local N = Int(length(q)/2)
```
Comment (PerezHz) [inline test/TestPkg/src/TestPkg.jl:132]: Why were the `local` moved inside the `for` loop?

**Review by PerezHz — APPROVED** [review]
Really, thank you for this contribution! I just left a couple of comments. Since tests are passing I would propose to merge this and tag a new minor version. Should any fix be needed, we can tackle that in another PR.

File: docs/src/taylorize.md [hunk 2]
```diff
@@ -16,17 +16,20 @@ repeatedly the function where the ODEs of the problem are defined, in
...
+performs better.
 The new method is constructed specifically for the actual function
-defining the equations of motion by parsing its expression; the new
-function performs in principle *exactly* the same operations, but avoids
-the extra allocations and the repetition of some operations.
+defining the equations of motion by parsing its expression. This new
+method performs in principle *exactly* the same operations, but avoids
+the extra allocations; to do the latter, the macro also creates an *internal*
+function `TaylorIntegration._allocate_jetcoeffs!`, which allocates all temporary
+`Taylor1` objects as well as the declared `Array{Taylo1,1}`s.
```
[the remark this answers is not in the document]
Comment (lbenet) [reply docs/src/taylorize.md:32]: Good catch! Fixed (in the forthcoming commit)!

File: src/explicitode.jl [hunk 3]
```diff
@@ -88,20 +88,28 @@ end
 
 # __jetcoeffs
 """
-    __jetcoeffs!(::Val{bool}, f, t, x, params)
-    __jetcoeffs!(::Val{bool}, f, t, x, dx, xaux, params)
+    __jetcoeffs!(::Val{bool}, f, t, x, params, , tmpTaylor, arrTaylor)
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/explicitode.jl:91]: Fixed!

File: test/TestPkg/src/TestPkg.jl (hunk 1 above)
Comment (lbenet) [reply test/TestPkg/src/TestPkg.jl:132]: I don't [see](https://github.com/PerezHz/TaylorIntegration.jl/blob/628abaf817e26583fcdd6d76e518d8c9bb1b19c9/test/TestPkg/src/TestPkg.jl#L19) that the `local` statements have been moved inside the `for` loop... Maybe changing the indentation led to this interpretation...

**Comment by lbenet** [comment]
@PerezHz There seems to be a problem with two log-checks for Julia v1.0.5: see [this](https://github.com/PerezHz/TaylorIntegration.jl/runs/6066211348?check_suite_focus=true#step:6:40) and [this](https://github.com/PerezHz/TaylorIntegration.jl/runs/6066211348?check_suite_focus=true#step:6:56). 

There are two work arounds:
- try to avoid the log-check check for `VERSION < v"1.6"`;
- require Julia v1.6 at least.

Both options are easy to implement. Yet, for the first one, I'm not sure if the problem will be solved, because it is related with the common interface (DiffEqBase and OrdinaryDiffEq), and there are some warnings about deprecations. (I don't have Julia 1.0 anymore in my machine, so I can't do the local tests.) The second option breaks (Julia) backward use. Once said that, [the current LTS version](https://julialang.org/downloads/#long_term_support_release) is v1.6. In any case, we should probably test explicitly the LTS version.

Any thoughts about this?

**Comment by lbenet** [comment]
Checking the first option...

**Comment by lbenet** [comment]
@PerezHz Tests are passing, including those for Julia v1.0 and v1.6. I think this is ready to be merged.

**Comment by PerezHz** [comment]
It's great that tests pass now on Julia 1.0, thanks for fixing that! Even so, eventually it might make sense to move from Julia v1.0, but we can leave that for another PR.

**Pull request merged**